Add a paying_for_college app

higs4281 opened the pull request:
This introduces an app, `paying_for_college`, that will provide a home for two current satellite apps: college-costs and comparisontool. This is part of a project to expand the use of the dynamic disclosures pages (run by college-costs) and to bring the comparisontool pages into Wagtail. 

We're starting with the the college-cost code, which was updated from Python 2.6 to Python 3.6 and can be brought into cfgov-refresh with full Python 3 test coverage.

This code can allow us to deprecate the college-costs repo and serve the current dynamic disclosures pages from our main repo. This is running currently on the DEV5 server for testing.
 
- Example disclosure on DEV5: 
`[DEV5]/paying-for-college2/understanding-your-financial-aid-offer/offer/?iped=459259&pid=5269&oid=0ED05E24A400B57774DECAFDCDFAE5BD197D8BE3&book=1503&gib=0&gpl=0&hous=10224&insi=0&insl=5020&inst=18&leng=0&mta=0&othg=8380&othr=6309&parl=0&pelg=0&perl=0&ppl=0&prvl=0&prvf=0&prvi=0&schg=0&stag=0&subl=0&totl=16205&tran=2340&tuit=8040&unsl=0&wkst=0`
- About-this-tool:  
`[DEV5]/paying-for-college2/understanding-your-financial-aid-offer/about-this-tool/`

- I've confirmed that the feedback page works and captures user entries.

Those are the only pages served by this code.

## Notes
- Most of this PR consists of porting our live college-costs code, minus a bunch of unneeded data files and templates. The app's namespace and db tables are unchanged.
- The intention of this first step is to get dynamic disclosures running in cfgov-refresh. To that end, the front-end disclosures assets are simply included in the new app's `static` folder.
- The next step would be moving comparisontool pages into Wagtail and making use of the more up-to-date `paying_for_college` data.

Fun fact: Ditching the standalone fixtures and skipping the college-cost wheel trimmed 7M off the final cfgov build.zip file.

chosak commented:
I love this idea. One concern is that recent college-costs frontend PRs (like [this one](https://github.com/cfpb/college-costs/pull/354) from an hour ago) need to be included in the static assets, as this PR will get out of date as changes are made over there. 

Would we then presumably migrate the frontend build here into cfgov-refresh in a separate PR?

higs4281 commented:
> recent college-costs frontend PRs

I've been monitoring those changes, and the one cited doesn't affect the assets here. In fact, most of the updated templates aren't even used (I've pointe this out)  and are not included here.

chosak commented on cfgov/paying_for_college/management/commands/update_pfc_national_stats.py:
```diff
@@ -0,0 +1,17 @@
+
+COMMAND_HELP = """update_pfc_national_stats gets the latest national statistics
+ yaml file from collegescorecard, parses it and updates our local json file at
+ college/fixtures/national_stats.json.
```
Is this another missed paying_for_college -> college -> paying_for_college un-renaming?

higs4281 replied:
Indeed it is.

chosak commented on cfgov/paying_for_college/models/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# flake8: noqa F401
+
+from django.conf import settings
```
Is this import needed?

higs4281 replied:
It is not.

chosak commented on cfgov/paying_for_college/models/disclosures.py:
```diff
@@ -0,0 +1,771 @@
+        ordering = ['nickname']
+
+
+class BAHRate(models.Model):
```
Is this BAHRate model needed/used anywhere? It doesn't seem to really be used/exposed [in college-costs](https://github.com/cfpb/college-costs/search?utf8=%E2%9C%93&q=bahrate&type=).

higs4281 replied:
Basic Allowance for Housing is not used in the current tool, but it is not yet clear whether it will  come into play for the public tool.

chosak commented on cfgov/paying_for_college/migrations/0001_initial.py:
```diff
@@ -0,0 +1,215 @@
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('contacts', models.TextField(blank=True, help_text='COMMA-SEPARATED LIST OF EMAILS')),
+                ('endpoint', models.TextField(blank=True)),
```
The equivalent migration in college-costs [here](https://github.com/cfpb/college-costs/blob/a75769f0df9d1ceabaf991cf83e4dba8b48c308b/paying_for_college/migrations/0001_initial.py#L65) has a limit of `max_length=255` for the `endpoint` and `name` fields because of the definitions [here](https://github.com/cfpb/college-costs/blob/54004216e4dd0fc2c460858b1e2689eb29fd9676/paying_for_college/models.py#L179-L180), which is changed in this PR [here](https://github.com/cfpb/cfgov-refresh/pull/5065/files#diff-7463f0a6105cc526d69708a8635bc723R179). Was this a deliberate change?

This will cause a problem if data greater than 255 characters is ever entered into these fields, because the database table is defined with that limit:

```sql
cfgov=> \d paying_for_college_contact;
                                        Table "cfpb.paying_for_college_contact"
    Column     |          Type          | Collation | Nullable |                        Default                         
---------------+------------------------+-----------+----------+--------------------------------------------------------
 id            | bigint                 |           | not null | nextval('paying_for_college_contact_id_seq'::regclass)
 endpoint      | character varying(255) |           | not null | 
 name          | character varying(255) |           | not null | 
 internal_note | text                   |           | not null | 
 contacts      | text                   |           | not null | 
Indexes:
    "idx_16899_primary" PRIMARY KEY, btree (id)
```

If we want to change these fields to remove the length restriction, there needs to be another Django migration to make the DB schema change.

higs4281 replied:
It was a deliberate change ... when I was planning to create new tables. As there is no longer an advantage to specifying varchar(255) in postgres, I was going to leave that behind, but as we're reusing the pfc tables, I should leave the definition as-is. Thanks for catching this.

chosak commented on cfgov/paying_for_college/models/disclosures.py:
```diff
@@ -0,0 +1,771 @@
+    contacts = models.TextField(
+        help_text="COMMA-SEPARATED LIST OF EMAILS",
+        blank=True)
+    endpoint = models.TextField(blank=True)
```
See comment above about removing the `max_length` restriction on the `endpoint` and `name` fields here.

higs4281 replied:
Model and migration restored to varchar.

chosak commented on cfgov/paying_for_college/search_indexes.py:
```diff
@@ -0,0 +1,28 @@
+class SchoolIndex(indexes.SearchIndex, indexes.Indexable):
+    text = indexes.CharField(document=True, model_attr='primary_alias')
+    school_id = indexes.IntegerField(model_attr='school_id')
+    nicknames = indexes.CharField()
```
I'm not super familiar with Haystack index definition, but I notice some minor differences here [with college-costs](https://github.com/cfpb/college-costs/blob/54004216e4dd0fc2c460858b1e2689eb29fd9676/paying_for_college/search_indexes.py#L9):

```diff
9c9
<     nicknames = indexes.CharField()
---
>     nicknames = indexes.CharField(model_attr='nicknames')
20,22d19
<     def prepare_nicknames(self, obj):
<         return obj.nicknames
< 
```

Do these do the same thing?

higs4281 replied:
I'm so glad you raised this. I thought this was a hidden error, because 'nicknames' is a model property, not a field, and I thought Haystack's model_attr option required the value to be a db field, the way a Django query does. In fact, model_attr [handles callables](https://django-haystack.readthedocs.io/en/v2.4.1/searchfield_api.html?highlight=model_attr#SearchField.model_attr), so this fix is not needed!

chosak commented on cfgov/paying_for_college/static/paying_for_college/disclosures/static/css/main.ie.css:
```diff
@@ -0,0 +1 @@
+/*! normalize.css v2.1.3 | MIT License | git.io/normalize */@import url(//fast.fonts.net/t/1.css?apiType=css&projectid=44e8c964-4684-44c6-a6e3-3f3da8787b50);.a-legend,.nemo embed,.nemo img,.nemo legend,.nemo object,img{max-width:100%}.a-legend,legend{white-space:normal}.nemo blockquote,.nemo q,q{quotes:none}.nemo a,sub,sup{vertical-align:baseline}.a-btn,a{text-decoration:none}.nemo pre,pre{word-w ...
```
FWIW when I diff these CSS files with the ones in college-costs I see some differences; I'll defer to a FEWD to review further.

higs4281 replied:
We're building these assets now rather than relying on a static copies.

chosak commented on cfgov/paying_for_college/tests/test_validators.py:
```diff
@@ -0,0 +1,106 @@
+
+INVALID_UUID = """
+\r\n\r\nOMG CHECK OUT THIS TOTES LEGIT CFPB-APPROVED WEBSITE: /
+http://nyan.cat\r\n\r\nOh, and here is some Kafka:\r\n\r\nOne morning, /
```
🐛

higs4281 replied:
Farewell Kafka.

chosak commented on cfgov/paying_for_college/views.py:
```diff
@@ -0,0 +1,339 @@
+
+    def get_context_data(self, **kwargs):
+        context = super(BaseTemplateView, self).get_context_data(**kwargs)
+        context['base_template'] = 'front/base_update.html'
```
I imagine that this could maybe be cleaned up further in future since the base template and root are fixed now that there's no standalone mode.

higs4281 replied:
Yes indeed.

higs4281 commented on cfgov/cfgov/settings/base.py:
```diff
@@ -61,6 +61,7 @@
     'taggit',
     'wagtailinventory',
     'wagtailsharing',
+    'paying_for_college',
```
Agreed. It would be nice to have all the apps that are native to cfgov-refresh in one section, but that's or another PR.

higs4281 commented on cfgov/.coveragerc:
```diff
@@ -9,6 +9,7 @@ omit =
     *settings/*
     */ask_cfpb/scripts/*
     */ask_cfpb/search_indexes*
+    */paying_for_college/data_sources/*
```
I'll make the exception specific to bls_processing.py. The script will probably be dumped when the current project addresses BLS data.

Somewhat related, I moved a number of historic data files, including those for the closed Perkins loan program, to the internal dynamic-disclosures-documentation repo for posterity.

higs4281 commented on cfgov/paying_for_college/data_sources/bls_processing.py:
```diff
@@ -0,0 +1,138 @@
+
+import json
+
+from college.models import cdr
```
an excellent point!

higs4281 commented on cfgov/paying_for_college/disclosures/scripts/ping_edmc.py:
```diff
@@ -0,0 +1,56 @@
+BIN = "https://httpbin.org/"
+BINPOST = "https://httpbin.org/post"
+BINGET = "https://httpbin.org/get"
+RBIN = "http://requestb.in/1ak4sxc1"
```
Good catch. That was a requestbin URL we were using to try to see what origin IPs the original school was seeing when we sent them notifications. Our server gateways were making this difficult to know for certain. 

I'll remove the outdated RBIN reference, but we will need to revisit notification reception in the current project, so some of this code could come in handy.

higs4281 commented on cfgov/paying_for_college/disclosures/urls.py:
```diff
@@ -0,0 +1,57 @@
+        name='pfc-feedback'),
+
+    url(r'^about-this-tool/$',
+        BaseTemplateView.as_view(template_name="pfc_disclosure_technote.html"),
```
Agreed.

chosak commented:
One other thought: college-costs currently exposes its [gh-pages](https://github.com/cfpb/college-costs/tree/gh-pages) branch on GitHub pages: http://cfpb.github.io/college-costs/. Should this content be moved over to [the cfgov-refresh documentation](https://cfpb.github.io/cfgov-refresh/) as well, maybe under the "Public APIs and Data" sidebar heading?

higs4281 commented:
> One other thought: college-costs currently exposes its [gh-pages](https://github.com/cfpb/college-costs/tree/gh-pages) branch on GitHub pages: http://cfpb.github.io/college-costs/. Should this content be moved over to [the cfgov-refresh documentation](https://cfpb.github.io/cfgov-refresh/) as well, maybe under the "Public APIs and Data" sidebar heading?

I've been chewing on the best place for this. The gh pages link has been provided to several schools as our official published specs for using the tool, so I think we're obliged to keep the docs live where they are, or at least keep alive a page that sends schools to a new documentation home.

I think documentation for new versions of the tool should live in cfgov-refresh, in the spot you suggest.

higs4281 commented on cfgov/paying_for_college/forms.py:
```diff
@@ -0,0 +1,12 @@
+    message = forms.CharField(widget=forms.Textarea)
+
+
+class EmailForm(forms.Form):
```
Good catch. Unneeded.

higs4281 commented on cfgov/paying_for_college/disclosures/urls.py:
```diff
@@ -0,0 +1,57 @@
+    url(r'^api/school/(\d+)/$',
+        SchoolRepresentation.as_view(),
+        name='school-json'),
+]
```
Unused. It was an artifact from comparisontool.

higs4281 commented on cfgov/paying_for_college/fixtures/test_fixture.json:
```diff
@@ -0,0 +1,253 @@
+    },
+    "model": "paying_for_college.notification",
+    "pk": 1
+}
```
Intentional. Another comparisontool artifact.

higs4281 commented:
@chosak Thank you so much for the thorough and thoughtful review. You must be exhausted!

I've addressed your comments, and we're now compiling assets via the cfgov-refresh front-end build. Which means we have some js linting to do.
